src/librustc_middle/ty/print/pretty.rs

```diff
@@ -1874,14 +1910,40 @@ impl fmt::Debug for TraitRefPrintOnlyTraitPath<'tcx> {
 
 impl ty::TraitRef<'tcx> {
     pub fn print_only_trait_path(self) -> TraitRefPrintOnlyTraitPath<'tcx> {
-        TraitRefPrintOnlyTraitPath(self)
+        TraitRefPrintOnlyTraitPath(self, false)
+    }
+    pub fn print_only_trait_path_full(self) -> TraitRefPrintOnlyTraitPath<'tcx> {
+        TraitRefPrintOnlyTraitPath(self, true)
+    }
+}
+
+/// Wrapper type for `ty::TraitRef` to force printing the full path instead
+/// of a possibly unique short path.
+#[derive(Copy, Clone, TypeFoldable, Lift)]
+pub struct TraitRefPrintFullPath<'tcx>(ty::TraitRef<'tcx>);
+
+impl fmt::Debug for TraitRefPrintFullPath<'tcx> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+impl ty::TraitRef<'tcx> {
+    pub fn print_full_path(self) -> TraitRefPrintFullPath<'tcx> {
+        TraitRefPrintFullPath(self)
     }
 }
 
 impl ty::Binder<ty::TraitRef<'tcx>> {
     pub fn print_only_trait_path(self) -> ty::Binder<TraitRefPrintOnlyTraitPath<'tcx>> {
         self.map_bound(|tr| tr.print_only_trait_path())
     }
+    pub fn print_only_trait_path_full(self) -> ty::Binder<TraitRefPrintOnlyTraitPath<'tcx>> {
+        self.map_bound(|tr| tr.print_only_trait_path_full())
+    }
+    pub fn print_full_path(self) -> ty::Binder<TraitRefPrintFullPath<'tcx>> {
+        self.map_bound(|tr| tr.print_full_path())
+    }
```

Comment:
These changes shouldn't be necessary once we stop using string matching for diagnostics, AFAICT.

Reply:
We still need to have access to the full path for structured suggestions.

Reply:
Presumably that can be handled by "just" using the "include `crate`" mode that these internally use, directly?